lnwire/dyn_propose.go

```diff
@@ -0,0 +1,249 @@
+package lnwire
+
+import (
+	"bytes"
+	"io"
+
+	"github.com/btcsuite/btcd/btcec/v2"
+	"github.com/btcsuite/btcd/btcutil"
+	"github.com/lightningnetwork/lnd/tlv"
+)
+
+const (
+	// DPDustLimitSatoshis is the TLV type number that identifies the record
+	// for DynPropose.DustLimit.
+	DPDustLimitSatoshis tlv.Type = 0
+
+	// DPMaxHtlcValueInFlightMsat is the TLV type number that identifies the
+	// record for DynPropose.MaxValueInFlight.
+	DPMaxHtlcValueInFlightMsat tlv.Type = 2
+
+	// DPChannelReserveSatoshis is the TLV type number that identifies the
+	// for DynPropose.ChannelReserve.
+	DPChannelReserveSatoshis tlv.Type = 4
+
+	// DPHtlcMinimumMsat is the TLV type number that identifies the record
+	// for DynPropose.HtlcMinimum.
+	DPHtlcMinimumMsat tlv.Type = 6
+
+	// DPToSelfDelay is the TLV type number that identifies the record for
+	// DynPropose.CsvDelay.
+	DPToSelfDelay tlv.Type = 8
+
+	// DPMaxAcceptedHtlcs is the TLV type number that identifies the record
+	// for DynPropose.MaxAcceptedHTLCs.
+	DPMaxAcceptedHtlcs tlv.Type = 10
+
+	// DPFundingPubkey is the TLV type number that identifies the record for
+	// DynPropose.FundingKey.
+	DPFundingPubkey tlv.Type = 12
+
+	// DPChannelType is the TLV type number that identifies the record for
+	// DynPropose.ChannelType.
+	DPChannelType tlv.Type = 14
+)
+
+// DynPropose is a message that is sent during a dynamic commitments negotiation
+// process. It is sent by both parties to propose new channel parameters.
+type DynPropose struct {
+	// ChanID identifies the channel whose parameters we are trying to
+	// re-negotiate.
+	ChanID ChannelID
+
+	// DustLimit, if not nil, proposes a change to the dust_limit_satoshis
+	// for the sender's commitment transaction.
+	DustLimit *btcutil.Amount
```

Comment:
I don't think these should be pointers. If we make the code correct to begin with, then the default initialization won't happen. Instead we could write unit tests to ensure the nothing gets default initialized

Reply:
See my other comment but one of the strongest beliefs I have about writing robust code is making the types as close to correct-by-construction as is possible, to minimize the number of invalid states. The pointers here are really a poor-mans Option[T] type. So I can replace the pointers with a real Option[T] but I strongly oppose just putting an undifferentiated buffer into this type where the collection of TLVs is.

Reply:
Np, with me re trying out `Option[T]` here if you so wish. It is the case that we need to distinguish between these values not being specified (no change needed) vs actually set. With the pointer, we'll need to always mind to make sure a deref doesn't happen. Also same goes when deciding if things should be written out in the first place: if it isn't specified, then we don't want to right a default value, instead we shouldn't write it at all.